SCClassLibrary/Common/Math/SimpleNumber.sc

```diff
@@ -705,13 +711,13 @@ SimpleNumber : Number {
 		};
 		hours = (decimal.div(3600) % 24).asString.padLeft(2, "0").add($:);
 		minutes = (decimal.div(60) % 60).asString.padLeft(2, "0").add($:);
-		seconds = (decimal % 60).asString.padLeft(2, "0").add($.);
-
-		// min value of precision is 0.001; this ensures that we stick to 3 decimal places in the
-		// formatted string.
-		precision = max(precision, 0.001);
-		mseconds = this.frac.round(precision) * 1000;
-		mseconds = mseconds.round.asInteger.asString.padLeft(3, "0");
+		seconds = (decimal % 60).asString.padLeft(2, "0");
+		if(decimalPlaces > 0, {
+			// this could be simplified once sclang gains sprintf functionality, see issue #3570
+			mseconds = "." ++ (number.frac.round(precision) * pow(10, decimalPlaces)).asInteger.asString.padLeft(decimalPlaces, "0");
```

Comment:
this will result in bad output if you pass in `decimalPlaces` > 11 or 12, because you'll overflow a 32-bit integer:

```supercollider
0.1.asTimeString(precision: 0.001, decimalPlaces: 13)
// -> 00:00:00.00-2147483648
```

a safer way to do it is accumulate a digit at a time while multiplying the fractional part by 10, although i believe you may lose precision that way. i'm not very familiar with numerical analysis, so i can't say for certain what the best way to approach it would be. perhaps it's not very important here anyway. but this function shouldn't produce garbage on valid inputs.

Reply:
here's something similar i've been using and that doesn't overflow...
```
(
~fixDec= {|val, numDecimals= 2|  //float to string with fixed number of decimals
        var str= val.round(0.1**numDecimals).asString;
        var num= str.size-str.indexOf($.)-1;
        str.extend(str.size+numDecimals-num, $0);
};
)

a= 0.1.rand
~fixDec.value(a, 13)
```

Reply:
thanks @redFrik ! but it fails on inputs that result in scientific notation: `~fixDec.value(1e-5, 13`